test/FunctionalTests/Microsoft.OData.Client.Tests/ALinq/SequenceMethodsTests.cs

```diff
@@ -0,0 +1,290 @@
+            }
+            catch (InvalidOperationException)
+            {
+                expectedExceptionThrown = true;
+            }
+            Assert.True(expectedExceptionThrown);
+        }
+
+        [Fact]
+        public void FirstOrDefaultPredicate()
+        {
+            // Test behaviour with one match:
+            _ctx.InterceptRequestAndAssertUri("/Customers?$filter=ContactName ne 'John'&$top=1");
+            _ctx.InterceptRequestAndMockResponseValue("Customers", "[" + TestContext.MockCustomer1 + "]");
+            Customer customer = _ctx.Customers.FirstOrDefault(c => c.Name != "John");
+            Assert.Equal("ALFKI", customer.Id);
+            Assert.Equal("Maria Anders", customer.Name);
+            Assert.Equal("Berlin", customer.City);
+
+            // Test behaviour with no match:
+            _ctx.InterceptRequestAndAssertUri("/Customers?$filter=ContactName eq 'John'&$top=1");
+            _ctx.InterceptRequestAndMockResponseValue("Customers", "[]");
+            Assert.Null(_ctx.Customers.FirstOrDefault(c => c.Name == "John"));
+        }
+
+        [Fact]
+        public void SinglePredicate()
+        {
+            // Test behaviour with one match:
+            _ctx.InterceptRequestAndAssertUri("/Customers?$filter=CustomerID eq 'CHOPS'&$top=2");
+            _ctx.InterceptRequestAndMockResponseValue("Customers", "[" + TestContext.MockCustomer2 + "]");
+            Customer customer = _ctx.Customers.Single(c => c.Id == "CHOPS");
+            Assert.Equal("CHOPS", customer.Id);
+            Assert.Equal("Yang Wang", customer.Name);
+            Assert.Equal("Bern", customer.City);
+
+            bool expectedExceptionThrown;
+
+            // Test behaviour with no match:
+            try
+            {
+                expectedExceptionThrown = false;
+                _ctx.InterceptRequestAndAssertUri("/Customers?$filter=ContactName eq 'thisdoesntexist'&$top=2");
+                _ctx.InterceptRequestAndMockResponseValue("Customers", "[]");
+                _ctx.Customers.Single(c => c.Name == "thisdoesntexist");
+            }
+            catch (InvalidOperationException)
+            {
+                expectedExceptionThrown = true;
+            }
+            Assert.True(expectedExceptionThrown);
+
+            // Test behaviour with more than one match:
+            try
+            {
+                expectedExceptionThrown = false;
+                _ctx.InterceptRequestAndAssertUri("/Customers?$filter=ContactName ne 'thisdoesntexist'&$top=2");
+                _ctx.InterceptRequestAndMockResponseValue("Customers", "[" + TestContext.MockCustomer1 + "," + TestContext.MockCustomer2 + "]");
+                _ctx.Customers.Single(c => c.Name != "thisdoesntexist");
+            }
+            catch (InvalidOperationException)
+            {
+                expectedExceptionThrown = true;
+            }
+            Assert.True(expectedExceptionThrown);
```

Comment:
Nitpick: Consider refactoring these out into separate methods. Also consider `Assert.Throws` to test exceptions. See: https://github.com/OData/odata.net/pull/2863/files#r1618197588